File: leetcode-python/_20xx/_2115_find_all_possible_recipes_from_supplies.py
import unittest
from typing import List

#2025-21-03
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        graph = {}
        for i, r in enumerate(recipes):
            graph[r] = ingredients[i]

        can_cooked = set(supplies)
        visit = set()
        def dfs(r: str) -> bool:
            if r in can_cooked:
                return True
            if r not in graph or r in visit:
                return False

            visit.add(r)
            for ingredient in graph[r]:
                if not dfs(ingredient):
                    return False
            can_cooked.add(r)
            return True

        ans = []
        for r in recipes:
            if dfs(r):
                ans.append(r)

        return ans
class MyTestCase(unittest.TestCase):

    def setUp(self) -> None:
        self.s = Solution()
    
    def test_case_1(self):
        actual = self.s.findAllRecipes(recipes = ["bread"], ingredients = [["yeast","flour"]], supplies = ["yeast","flour","corn"])
        self.assertEqual(["bread"], actual)
        
    def test_case_2(self):
        actual = self.s.findAllRecipes(recipes = ["bread","sandwich"], ingredients = [["yeast","flour"],["bread","meat"]], supplies = ["yeast","flour","meat"])
        self.assertEqual(["bread","sandwich"], actual)

    def test_case_3(self):
        actual = self.s.findAllRecipes(recipes = ["bread","sandwich","burger"],
                                       ingredients = [["yeast","flour"],["bread","meat"],["sandwich","meat","bread"]],
                                       supplies = ["yeast","flour","meat"])
        self.assertEqual(["bread","sandwich","burger"], actual)


    def test_case_4(self):
        actual = self.s.findAllRecipes(recipes = ["B","A","C"],
                                       ingredients = [["D","E"],["B","F"],["B","G"]],
                                       supplies = ["D","E","F"])
        self.assertEqual(["B","A"], actual) # No G supplies

    def test_case_5(self):
        actual = self.s.findAllRecipes(recipes=["ju","fzjnm","x","e","zpmcz","h","q"],
                                       ingredients=[
                                           ["d"],
                                           ["hveml","f","cpivl"],
                                           ["cpivl","zpmcz","h","e","fzjnm","ju"],
                                           ["cpivl","hveml","zpmcz","ju","h"],
                                           ["h","fzjnm","e","q","x"],
                                           ["d","hveml","cpivl","q","zpmcz","ju","e","x"],
                                           ["f","hveml","cpivl"]
                                       ],
                                       supplies=["f","hveml","cpivl","d"])
        self.assertEqual(["ju", "fzjnm", "q"], actual)  # No G supplies

    def test_case_6(self):
        actual = self.s.findAllRecipes(recipes=["burger","bread","sandwich"],
                                       ingredients=[["sandwich","meat","bread"],["yeast","flour"],["bread","meat"]],
                                       supplies=["yeast","flour","meat"])
        self.assertEqual(["burger","bread","sandwich"], actual)  # No G supplies
if __name__ == '__main__':
    unittest.main()

File: leetcode-cpp/_20xx/_2115_find_all_possible_recipes_from_supplies.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>

#include "../test/catch_amalgamated.hpp"

using namespace std;

class Solution {
   public:
    vector<string> findAllRecipes(vector<string> &recipes,
                                  vector<vector<string>> &ingredients,
                                  vector<string> &supplies) {
        auto graph = this->buildGraph(recipes, ingredients);
        auto canCooked = unordered_set(supplies.begin(), supplies.end());
        auto visit = unordered_set<string>{};

        auto ans = vector<string>{};
        for (auto r : recipes) {
            if (this->dfs(r, graph, canCooked, visit)) {
                ans.push_back(r);
            }
        }
        return ans;
    }

    bool dfs(string r, unordered_map<string, vector<string>> graph,
             unordered_set<string> canCooked, unordered_set<string> visit) {
        if (canCooked.contains(r)) {
            return true;
        }
        if (visit.contains(r) || !graph.contains(r)) {
            return false;
        }
        visit.insert(r);
        auto neighbours = graph[r];
        for (auto ingredient : neighbours) {
            if (!this->dfs(ingredient, graph, canCooked, visit)) {
                return false;
            }
        }
        canCooked.insert(r);
        return true;
    }

    unordered_map<string, vector<string>> buildGraph(
        vector<string> &recipes, vector<vector<string>> &ingredients) {
        unordered_map<string, vector<string>> graph;
        for (size_t i = 0; i < recipes.size(); i++) {
            graph.insert(std::make_pair(recipes.at(i), ingredients.at(i)));
        }

        return graph;
    }
};

TEST_CASE("test_case_1") {
    Solution s;
    auto recipes = vector<string>{"bread"};
    auto ingredients = vector<vector<string>>{{"yeast", "flour"}};
    auto supplies = vector<string>{"yeast", "flour", "corn"};
    auto actual = s.findAllRecipes(recipes, ingredients, supplies);
    auto expect = vector<string>{"bread"};

    CHECK(actual == expect);
}

TEST_CASE("test_case_2") {
    Solution s;
    auto recipes = vector<string>{"bread", "sandwich"};
    auto ingredients =
        vector<vector<string>>{{"yeast", "flour"}, {"bread", "meat"}};
    auto supplies = vector<string>{"yeast", "flour", "meat"};
    auto actual = s.findAllRecipes(recipes, ingredients, supplies);
    auto expect = vector<string>{"bread", "sandwich"};
    CHECK(actual == expect);
}

TEST_CASE("test_case_3") {
    Solution s;
    auto recipes = vector<string>{"bread", "sandwich", "burger"};
    auto ingredients = vector<vector<string>>{
        {"yeast", "flour"}, {"bread", "meat"}, {"sandwich", "meat", "bread"}};
    auto supplies = vector<string>{"yeast", "flour", "meat"};
    auto actual = s.findAllRecipes(recipes, ingredients, supplies);
    auto expect = vector<string>{"bread", "sandwich", "burger"};
    CHECK(actual == expect);
}

TEST_CASE("test_case_5") {
    Solution s;
    auto recipes = vector<string>{"ju", "fzjnm", "x", "e", "zpmcz", "h", "q"};
    auto ingredients = vector<vector<string>>{
        {"d"},
        {"hveml", "f", "cpivl"},
        {"cpivl", "zpmcz", "h", "e", "fzjnm", "ju"},
        {"cpivl", "hveml", "zpmcz", "ju", "h"},
        {"h", "fzjnm", "e", "q", "x"},
        {"d", "hveml", "cpivl", "q", "zpmcz", "ju", "e", "x"},
        {"f", "hveml", "cpivl"}};
    auto supplies = vector<string>{"f", "hveml", "cpivl", "d"};
    auto actual = s.findAllRecipes(recipes, ingredients, supplies);
    auto expect = vector<string>{"ju", "fzjnm", "q"};
    CHECK(actual == expect);
}

TEST_CASE("test_case_6") {
    Solution s;
    auto recipes = vector<string>{"burger", "bread", "sandwich"};
    auto ingredients = vector<vector<string>>{
        {"sandwich", "meat", "bread"}, {"yeast", "flour"}, {"bread", "meat"}};
    auto supplies = vector<string>{"yeast", "flour", "meat"};
    auto actual = s.findAllRecipes(recipes, ingredients, supplies);
    auto expect = vector<string>{"burger", "bread", "sandwich"};
    CHECK(actual == expect);
}
